Fix kubectl sort-by metadata issue.

aveshagarwal opened the pull request:
Internal types are not supposed to have json metadata (though in kubernetes
they do) as it is true with openshift types. That means sort-by must work
on versioned objects for sorting, otherwise it produces "error: metadata
is not found" error if it sorts internal types without json metadata.

This PR converts internal types objects to versioned objects and sort-by
sorts them correctly without medata error, and then it prints
corresponding internal objects in sorted order.

aveshagarwal commented:
@kubernetes/rh-cluster-infra

ncdc commented:
cc @kubernetes/rh-ux @kubernetes/kubectl

liggitt commented on pkg/kubectl/cmd/get.go:
```diff
@@ -248,6 +248,22 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
+		}
+
 		// TODO: questionable
 		if sorter, err = kubectl.SortObjects(f.Decoder(true), objs, sorting); err != nil {
```
at first glance, rather than converting into versions objects, here, I *think* would expect to pass the output version to `SortObjects`... it is taking a fieldInput, which is not really meaningful unless paired with a group version. In `SortObjects`, can we decode to the specified groupversion? I'd like @smarterclayton's take on that as well...

smarterclayton replied:
I would expect us to be sorting based on the JSON and the field path
qualifier only, or at minimum on the map[string]interface{}.  It's a
riskier change though.

On Mon, Feb 22, 2016 at 1:37 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/kubectl/cmd/get.go
> <https://github.com/kubernetes/kubernetes/pull/21694#discussion_r53669745>
> :
>
> > +
> > +		clientConfig, err := f.ClientConfig()
> > +		if err != nil {
> > +			return err
> > +		}
> > +
> > +		version, err := cmdutil.OutputVersion(cmd, clientConfig.GroupVersion)
> > +		if err != nil {
> > +			return err
> > +		}
> > +
> > +		objs, err = resource.AsVersionedObjects(infos, version.String(), f.JSONEncoder())
> > +		if err != nil {
> > +			return err
> > +		}
> > +
> >  		// TODO: questionable
> >  		if sorter, err = kubectl.SortObjects(f.Decoder(true), objs, sorting); err != nil {
>
> at first glance, rather than converting into versions objects, here, I
> *think* would expect to pass the output version to SortObjects... it is
> taking a fieldInput, which is not really meaningful unless paired with a
> group version. In SortObjects, can we decode to the specified
> groupversion? I'd like @smarterclayton <https://github.com/smarterclayton>'s
> take on that as well...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/21694/files#r53669745>.
>

aveshagarwal replied:
@liggitt I am looking into what you suggested and i think it should be possible. The only issue with that I see is that once we decode objects into specified groupversion inside SortObjects, after sorting, any caller of sort-by would have converted versioned objects and it would require handling printing of those objects in the similar way as it is done currently in kubectl get. Because as I understand, it is not possible to print versioned objects directly but only internal types objects.

aveshagarwal replied:
Or another way might be to convert sorted and decoded obejcts back to their internal types and then return them. Although I am not sure how to handle this,

ncdc commented on pkg/kubectl/cmd/get.go:
```diff
@@ -248,6 +248,22 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 	sorting, err := cmd.Flags().GetString("sort-by")
 	var sorter *kubectl.RuntimeSort
 	if err == nil && len(sorting) > 0 && len(objs) > 1 {
+
```
Remove blank line

ncdc commented:
e2e failed because of

```
Error pulling image (0.3) from gcr.io/google_containers/kube-registry-proxy, endpoint: https://gcr.io/v1/, Get https://storage.googleapis.com/artifacts.google-containers.appspot.com/containers/images/9b9342c134bbda6f57c95c52cafd4ae40333dd6cc0fce4bace9296f957042d22/ancestry: dial tcp 74.125.202.128:443: i/o timeout
```

ncdc commented:
@k8s-bot test this issue: #IGNORE

aveshagarwal commented:
I have updated this PR and it seems to be working wth kubectl get. I am still trying to see if it could affect annotate, create, rollingupdate, expose, run, autoscale, label,  although I have yet to check if these commands really need sort-by or not.

liggitt commented:
Wouldn't anything that outputs objects potentially use SortObjects and be affected? That's why I think the work of applying the sort by field to the correct output version should be done inside the common sorting code

kargakis commented:
I think sort-by was added only for `kubectl get` and no other command really uses it.

liggitt commented:
is it a global flag? If so, we should stop adding global flags that only apply to a single command

lavalamp commented on pkg/kubectl/sorting_printer.go:
```diff
@@ -63,7 +66,7 @@ func (s *SortingPrinter) sortObj(obj runtime.Object) error {
 	}
 
-	sorter, err := SortObjects(s.Decoder, objs, s.SortField)
+	sorter, err := SortObjects(s.Decoder, objs, s.SortField, "")
```
We should probably never sort by the "" version or group

lavalamp commented:
Wouldn't it be a lot easier to just convert to the output version to begin with, and then nothing else would need to change?

kargakis commented:
> is it a global flag? If so, we should stop adding global flags that only apply to a single command

Agreed

aveshagarwal commented:
I have updated this PR based on previous feedback and also did some refactoring.

aveshagarwal commented:
Fixed unit tests.

aveshagarwal commented:
I am still looking into unit and e2e failures.

liggitt commented on pkg/kubectl/cmd/util/factory.go:
```diff
@@ -838,9 +842,9 @@ func (f *Factory) PrinterForMapping(cmd *cobra.Command, mapping *meta.RESTMappin
-		printer = maybeWrapSortingPrinter(cmd, printer)
 	}
 
+	printer = maybeWrapSortingPrinter(cmd, printer, version)
```
what does this do if we didn't get in the `ok` block and `version` is undefined?

liggitt commented on pkg/kubectl/sorting_printer.go:
```diff
@@ -80,7 +84,7 @@ func (s *SortingPrinter) sortObj(obj runtime.Object) error {
 }
 
-func SortObjects(decoder runtime.Decoder, objs []runtime.Object, fieldInput string) (*RuntimeSort, error) {
+func SortObjects(decoder runtime.Decoder, objs []runtime.Object, fieldInput string, version string) (*RuntimeSort, error) {
```
this should be an actual groupversion object... no reason to lose the structure

liggitt commented on pkg/kubectl/sorting_printer.go:
```diff
@@ -93,13 +97,20 @@ func SortObjects(decoder runtime.Decoder, objs []runtime.Object, fieldInput stri
+		var err error
+		partialType := strings.Split(reflect.TypeOf(objs[ix]).String(), ".")[0]
+		switch partialType {
+		case "*api":
```
*raises eyebrow*

kargakis replied:
:)

liggitt replied:
this seems odd... why wouldn't we always attempt to convert to the output version for non *runtime.Unknown objects, and in error cases, just use the object as-is for sorting?

Commit:
Fix kubectl sort-by metadata issue.

Internal types are not supposed to have json metadata (though in kubernetes
they do) as it is true with openshift types. That means sort-by must work
on versioned objects for sorting, otherwise it produces "error: metadata
is not found" error if it sorts internal types without json metadata.

This PR converts internal types objects to versioned objects and sort-by
sorts them correctly without medata error, and then it prints
corresponding internal objects in sorted order.

aveshagarwal commented:
I have updated this PR and now kubectl get passes versioned objects to sort-by. It seems reasonable to me that sort-by should just focus on sorting not object conversion.

lavalamp commented on pkg/kubectl/cmd/get.go:
```diff
@@ -262,10 +279,13 @@ func RunGet(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string
 
 	for ix := range objs {
 		var mapping *meta.RESTMapping
+		var original runtime.Object
```
Is this still needed? Don't you want to print out a versioned object anyway?

aveshagarwal replied:
How to print versioned objects? Whenever I try to print versioned objects, for example v1, I get 

```
error: unknown type &v1.Pod{TypeMeta:unversioned.TypeMeta{Kind:"Pod", APIVersion:"v1"},....
```
Complete error is here: http://fpaste.org/331443/67944711/

aveshagarwal replied:
And the reason seems to be that all object printers in pkg/kubectl/resource_printer.c seem to be operating on internal types unless I am missing something.

liggitt replied:
@lavalamp all the registered column printers and describers work from internal types.

lavalamp replied:
OK

aveshagarwal commented:
e2e flake:
```
[Fail] Deployment [It] deployment should delete old replica sets 
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/deployment.go:440
```

aveshagarwal commented:
@k8s-bot test this github issue https://github.com/kubernetes/kubernetes/issues/22088

liggitt commented:
this is obscure enough that it deserves at least one test, I think

aveshagarwal commented:
@liggitt ok will look into that.

aveshagarwal commented:
@liggitt  Added a test case. PTAL.

liggitt commented:
LGTM
